man/man8/dbmbuild.8
.TH DBMBUILD 8
.\" @(#) $Header: /xtel/pp/pp-beta/man/man8/RCS/dbmbuild.8,v 6.0 1991/12/18 20:44:03 jpo Rel $
.\"
.\" $Log: dbmbuild.8,v $
.\" Revision 6.0  1991/12/18  20:44:03  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
dbmbuild \- database builder for PP
.SH SYNOPSIS
.in +.5i
.ti -.5i
dbmbuild \%[-nvdk] \%[database\ [table ...]] 
.in -.5i
.SH DESCRIPTION
.I Dbmbuild
reads the tables specified in the PP tailor file into a hashed
database for use in quickly verifying addresses and efficiently assigning
channels to submitted messages.  If no database file is specified, the
default dbm filename is used from the PP tailor file.  If no table files
are specified, all tables listed in the tailor file are used. 
.PP
The options are:
.IP n
Don't create a new database, attempt to add to an existing database.
.IP v
Verbose mode.  Print out each table as it is processed.
.IP d
Debug mode.  Report everything that happens (verbose).
.IP k
Keep going.  If a file is mentioned that doesn't exist, ignore it.
It might be appropriate to make this the default at some sites.
.PP
Appropriate locks are placed on the database so that
.I dbmbuild
can safely be run while PP is in operation.
.SH FILES
.nf
.ta \w'ppdbm.dir\0\0'u
ppdbm.dir	database directory
ppdbm.pag	database pages
ppdbm.lck	database locking file
.fi
.SH "SEE ALSO"
dbm(3X), dbmedit(8), dbmunbuild(8)
.br
\fIThe PP Manual: Volume 1 \- Installation and Operation\fP


CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Tools/dbmbuild/dbmbuild.c
/* dbmbuild: build the database of information */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Tools/dbmbuild/RCS/dbmbuild.c,v 6.0 1991/12/18 20:29:34 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Tools/dbmbuild/RCS/dbmbuild.c,v 6.0 1991/12/18 20:29:34 jpo Rel $
 *
 * $Log: dbmbuild.c,v $
 * Revision 6.0  1991/12/18  20:29:34  jpo
 * Release 6.0
 *
 */



/* ---------------------------------------------------------------------------


	Database builder for PP address database

	Command line:
		dbmbuild [-flags] [outfile]

		where flags are:
		d   -   debugging output
		v   -   verbose output
		k   -   keep going if a table is missing
		n   -   don't make a new database file

	output-file is name of EXISTING dbm file to insert entires.
	Note it is two files, an output-file.dir and an output-file.pag.

--------------------------------------------------------------------------- */


#include        "head.h"
#include        "util.h"
#include        "chan.h"
#include        "dbase.h"
#include 	"sys.file.h"
#include        <varargs.h>

extern int      errno;
extern char     *tbldfldir,
		*ppdbm;

#ifdef GDBM
GDBM_FILE	thedb;
#define store(x,y)	gdbm_store (thedb, x, y, GDBM_REPLACE)
#define fetch(x)	gdbm_fetch (thedb, x)
#else
#ifndef GCC_DBM_OK
#if sparc && defined(__GNUC__)	/* work around bug in gcc 1.37 sparc version */
 #error GCC and dbm do not get along on a sparc - compile with cc.
#endif
#endif
#endif

#ifdef NDBM
DBM             *thedb;
#define         store(x,y)      dbm_store (thedb, (x), (y), DBM_REPLACE)
#define         fetch(x)        dbm_fetch (thedb, (x))
#endif


int             error,
		Debug,
		Verbose,
		newflag = 1,
		keepgoing;

char            dbfile [128],
		*dblock;

static char *myname;
static short    tb_nopen;   /* -- no of opened file descriptors -- */
void    adios (), advise ();

static int theinit (), theend ();
static int dbfinit (), dbfclose ();
static int process ();
static int install ();
static int prdatum ();
static int check ();
static int tb_open (), tb_close (), tb_free ();
static void cleanup ();

/* ---------------------  Begin  Routines  -------------------------------- */


/*
Process arguments, set flags and invoke file processing.
then clean up and exit properly.
*/


main (argc, argv)
char                    **argv;
{
	register Table      *tp;
	char                *p,
	*outfile = NULLCP;
	int                 ind;
	int		opt;
	extern char *optarg;
	extern int optind;

	myname = argv[0];
	sys_init (myname);

	while ((opt = getopt (argc, argv, "dknv")) != EOF) {
		switch (opt) {
		    case 'd':
			Debug++;
			break;
		    case 'k':
			keepgoing++;
			break;
		    case 'n':
			newflag = 0;
			break;
		    case 'v':
			Verbose++;
			break;
		    default:
			adios (NULLCP, "Usage: %s [-vnkd] [database]", myname);
		}
	}

	if (optind < argc)
		optarg = argv[optind++];

	/* -- use default database -- */

	if (outfile == NULLCP) {
		if (!isstr (ppdbm)) {
			adios (NULLCP,
			       "no default data base, in 'ppdbm' variable");
		}
		outfile = ppdbm;
	}


	/* -- check for existence first -- */

	error = 0;

	if (optind >= argc) {

		for (ind = 0; ch_all[ind] != NULLCHAN; ind++) {
			if (ch_all[ind] -> ch_table != NULLTBL)
				error += check (ch_all[ind] -> ch_table);

			if (Debug)
				fprintf (stderr, "ch_show = '%s' - %d\n",
					 ch_all[ind] -> ch_show, error);
		}


		for (ind = 0; (tp = tb_all[ind]) != NULLTBL; ind++) {
			error += check (tp);
			if (Debug)
				fprintf (stderr, "tb_show = '%s' - %d\n",
					 tp -> tb_show, error);
		}
	}
	else {
		int	ac = optind;
		while (ac < argc) {
			if ((tp = tb_nm2struct (argv[ac++])) == NULLTBL) {
				advise (NULLCP, "Table '%s' is unknown",
					argv[ac - 1]);
				error++;
			}
			error += check (tp);
		}
	}

	if (error)
		adios (NULLCP, "Some tables are missing.  dbmbuild aborted.");

	getfpath (tbldfldir, outfile, dbfile);

	if (Verbose || Debug) 
		fprintf (stderr, 
"database info:\n\ttables from %s\n\tdatabase = %s{.dir,.pag}\n\tpath = %s\n",
			 tbldfldir, outfile, dbfile);

	dblock = multcat (dbfile, ".lck", NULLCP);

	(void) close (creat (dblock, 0444 ));

	if (!theinit (newflag))
		cleanup (NOTOK);

	if (optind >= argc) {

		/* -- process for real! -- */

		for (ind = 0; ch_all[ind] != NULLCHAN; ind++)
			process (ch_all[ind] -> ch_table);


		for (ind = 0; (tp = tb_all[ind]) != NULLTBL; ind++) {

			/* -- in case table used by chan -- */

			if (tp -> tb_fp != (FILE *)EOF) {
				if (Debug)
					fprintf (stderr, "Table '%s'\n", tp -> tb_name);
				process (tp);
			}
		}

	}
	else
		while (optind < argc) {

			if ((tp = tb_nm2struct (argv[optind++])) == NULLTBL) {
				(void) fflush (stdout);
				adios (NULLCP, "Table '%s' is unknown",
				       argv[optind-1]);
			}
			process (tp);
		}

	cleanup (theend (newflag));
}


/* ---------------------  Static  Routines  ------------------------------- */




static int theinit (new)   /* -- initialize the dbm files -- */
int             new;
{
    char        tmpfile [100];


    /* -- start with fresh files -- */

    if (new) {
	if (Verbose || Debug)
	    fprintf (stderr, "Temporary database:  %s$\n", dbfile);
#ifdef DBM
	(void) sprintf (tmpfile, "%s$.pag", dbfile);

	if (Debug)
	    fprintf (stderr, "creating '%s'\n", tmpfile);


	if (close (creat (tmpfile, 0644)) < 0)
	    adios (tmpfile, "could not creat");

	(void) chmod (tmpfile, 0644);   /* -- in case umask screwed us -- */

	(void) sprintf (tmpfile, "%s$.dir", dbfile);

	if (Debug)
	    fprintf (stderr, "creating '%s'\n", tmpfile);


	/* -- create and/or zero the file -- */

	if (close (creat (tmpfile, 0644)) < 0)
	    adios (tmpfile, "could not creat");

	(void) chmod (tmpfile, 0644);   /* -- in case umask screwed us -- */
#endif
#ifdef GDBM
	(void) sprintf (tmpfile, "%s$.gdbm", dbfile);
#else
	(void) sprintf (tmpfile, "%s$", dbfile);
#endif
	return (dbfinit (tmpfile));
    }

    return (dbfinit (dbfile));
}




static theend (new)               /* cleanup the dbm files */
int             new;
{
    char        fromfile [100],
		tofile [100];


    dbfclose();

    /* -- started with fresh files -- */

    if (new) {
	if (Verbose || Debug)
	    fprintf (stderr, "Moving to database:  %s\n", dbfile);

#ifdef GDBM
	(void) sprintf (fromfile, "%s$.gdbm", dbfile);
	(void) sprintf (tofile, "%s.gdbm", dbfile);
	if (Debug)
		fprintf (stderr, "moving '%s'\n", fromfile);
	if (rename (fromfile, tofile) == NOTOK)
		adios (tofile, "could not rename %s to ", fromfile);
#else
	(void) sprintf (fromfile, "%s$.pag", dbfile);

	(void) sprintf (tofile, "%s.pag", dbfile);

	if (Debug)
	    fprintf (stderr, "moving '%s'\n", fromfile);

	(void) unlink (tofile);

	/* -- create and/or zero the file -- */

	if (rename (fromfile, tofile ) < 0 )
		adios (tofile, "could not rename %s to ", fromfile);

	(void) sprintf (fromfile, "%s$.dir", dbfile);

	(void) sprintf (tofile, "%s.dir", dbfile);

	if (Debug)
	    fprintf (stderr, "moving '%s'\n", fromfile);

	(void) unlink (tofile);

	/* -- create and/or zero the file -- */

	if (rename (fromfile, tofile ) == NOTOK)
		adios (tofile, "could not rename %s to ", fromfile);
#endif

	return (TRUE);
    }

    return (TRUE);
}




/*
Initialize the dbm file. Fetch the local name datum
Init to 1 and store it if there is no datum by that name.
*/

static dbfinit (filename)
char    *filename;
{
#ifdef GDBM
	if ((thedb = gdbm_open (filename, 0, GDBM_NEWDB, 0644, NULL)) == NULL)
#else
#ifdef NDBM
	if ((thedb = dbm_open (filename, O_CREAT|O_RDWR, 0644)) == NULL)
#else
	if (dbminit (filename) < 0)
#endif
#endif
	    adios (filename, "could not initialize data base");
	return (1);
}




/*
Close the dbm datafile.
Can't close the file because the library does not provide the call...
*/

static dbfclose()
{
#ifdef NDBM
	dbm_close (thedb);
#endif
#ifdef GDBM
	gdbm_close (thedb);
#endif
}




/*
Process a sequential file and insert items into the database
Opens argument assumes database is initialized.
*/

static process (tp)
Table           *tp;
{
    datum       key,
		value;
    char        tbkey [LINESIZE],
		tbvalue [LINESIZE],
		*cp;
    int status;

    if (tp == NULLTBL)
	return;

    if (Verbose || Debug)
	fprintf (stderr, tp -> tb_fp == (FILE *)EOF ?
			"%s (%s) already done\n" : "%s (%s)\n",
			tp -> tb_show, tp -> tb_name);

    if (tp -> tb_fp == (FILE *)EOF)
	return;


    /* -- gain access to a channel table -- */

    if (!tb_open (tp)) {
	fprintf (stderr, "could not open table \"%s\" (%s, file = '%s'):\n\t",
			tp -> tb_show, tp -> tb_name,
			tp -> tb_file);
	perror ("");

	/* -- don't try again -- */
	tp -> tb_fp = (FILE *)EOF;

	if (keepgoing)
	    return;
	cleanup (NOTOK);
    }


    tbkey[0] = tbvalue[0] = '\0';
    while ((status = tab_fetch (tp -> tb_fp, tbkey, tbvalue)) != DONE) {

	if (status == NOTOK) {
		if (!keepgoing)
			adios (NULLCP, "Bad record in file %s (last was %s:%s)",
				 tp -> tb_name, tbkey, tbvalue);
		advise (NULLCP, "Bad record in file %s (last was %s:%s)", 
				tp -> tb_name, tbkey, tbvalue);
		continue;
	}

	value.dptr = tbvalue;
	value.dsize = strlen (tbvalue) + 1;

	/* -- all keys are lower case -- */
	for (cp = tbkey; *cp != 0; cp++)
	    *cp = uptolow (*cp);

	key.dptr = tbkey;
	key.dsize = strlen (tbkey) + 1;

	if (Debug)
	    fprintf (stderr, " (%d)'%s': (%d)'%s'\n",
		     key.dsize, key.dptr, value.dsize, value.dptr);

	install (key, value, tp -> tb_name);
    }

    if (ferror (tp -> tb_fp))
	advise ("", "i/o error with table %s (%s, file = %s)",
			    tp -> tb_show, tp -> tb_name,
			    tp -> tb_file);

    (void) tb_close (tp);

    /* -- don't try again -- */
    tp -> tb_fp = (FILE *)EOF;
}




/*
Install a datum into the database.
Fetch entry first to see if we have to append name for building entry.
*/

static install (key, value, tbname)
datum           key,
		value;
char            tbname[];
{
    datum       old;
    char        newentry [ENTRYSIZE],
		*p,
		*q;

    p = newentry;

    old = fetch (key);

    if (old.dptr != NULLCP) {
	if (Debug) {
	    fprintf (stderr, "\tFound old entry\n\t");
	    prdatum (old);
	}
	for (p = newentry, q = old.dptr ; *p++ = *q++;);
	*(p-1) = FS;
    }
    else
	*p = '\0';

    (void) sprintf (p, "%s %s", tbname, value.dptr);

    old.dptr = newentry;
    old.dsize = strlen (newentry)+1;

    if (Debug) {
	fprintf (stderr, "\tNew datum\n\t");
	prdatum (old);
    }


    /* -- put the datum back -- */

    if (store (key, old) < 0) {
	advise ("",
		"store failed; key='%s'", key.dptr);
	prdatum (old);
	cleanup (NOTOK);
    }

}




/*
Print a datum.
Takes the datum arg & prints it so it can be read as either a key or entry.
*/

static prdatum (value)
datum           value;
{
	int     cnt;
	char    data [512];

	(void) strcpy (data, value.dptr);

	for (cnt = 0; cnt < value.dsize; cnt++) {
		if (value.dptr [cnt] >= ' ' && value.dptr [cnt] <= '~')
			putc (value.dptr [cnt], stderr);
		else
			fprintf (stderr, "<\\%03o>", value.dptr [cnt]);
	}

	putc ('\n', stderr);
}




static check (tp)
Table   *tp;
{

    /* -- its bad -- */
    if (tp == NULLTBL) {
	if (Debug)
		fprintf (stderr, "check's tp is NULL\n");
	return (1);
    }

    if (tp->tb_fp == (FILE *)EOF) {
	if (Debug)
		fprintf (stderr, "check's tp has a EOF FILE \n");
	return (1);
    }

    /* -- gain access to a channel table -- */
    if (!tb_open (tp)) {
	advise ("", "could not open table \"%s\" (%s, file = '%s')",
		tp -> tb_show, tp -> tb_name, tp -> tb_file);
	tp -> tb_fp = (FILE *)EOF;
	return (1);
    }

    (void) tb_close (tp);
    tp -> tb_fp = NULLFILE;
    return (0);
}




static void 	cleanup (exitval)
int     exitval;
{
	exit (exitval == TRUE? 0 : 1);  /* TRUE is non-zero */
}

/* -------------  Basic  File  Action  ------------------------------------ */

static tb_open (tp)
register Table  *tp;   /* -- table's state information -- */
{
    char        temppath [128];


    /*
    currently, there is one file (descriptor) per table.  for a large
    configuration, you will run out of fd's.  until/unless you run
    a single-file dbm-based version, there is a hack, below, which
    limits the number of open file-descriptors to 6.  Trying to open
    a 7th will cause a currently-opened one to be closed.
    */

    if (tp == NULLTBL) {
	if (Debug)
		fprintf (stderr, "tb_open's table is NULL\n");
	return (FALSE);
    }

    if (tb_nopen < 6 || tb_free()) {

	getfpath (tbldfldir, tp -> tb_file, temppath);

	/* -- add on directory portion -- */
	if ((tp -> tb_fp = fopen (temppath, "r")) != NULLFILE) {
		tb_nopen++;
		return (TRUE);
	}
    }

    printf ("cannot open %s\n", temppath);

    return (FALSE);
}




static tb_close (tp)
register Table  *tp;
{
	if (tp -> tb_fp == (FILE *)NOTOK || tp -> tb_fp == NULLFILE)
		return (FALSE);

	(void) fclose (tp -> tb_fp);

	/* -- mark as free -- */
	tp -> tb_fp = 0;

	tb_nopen--;

	return (TRUE);
}




static tb_free()   /* -- create a free file descriptor -- */
{
    register Table      **tpp;

    /*
    step from the lowest search priority to the highest, looking for a
    channel to close.  return success as soon as one is done.
    */

    for (tpp = tb_all; *tpp != 0; tpp++) {
	if (tb_close (*tpp))
		return (TRUE);   /* -- got a channel closed -- */
    }
    return (FALSE);   /* -- couldn't get any closed -- */
}


#ifndef lint
static void    _advise ();


void    adios (va_alist)
va_dcl
{
	va_list ap;

	va_start (ap);

	_advise (ap);

	va_end (ap);

	_exit (1);
}
#else
/* VARARGS */

void    adios (what, fmt)
char   *what,
       *fmt;
{
	adios (what, fmt);
}
#endif


#ifndef lint

void    advise (va_alist)
va_dcl
{
	va_list ap;

	va_start (ap);

	_advise (ap);

	va_end (ap);
}

static void  _advise (ap)
va_list ap;
{
	char    buffer[BUFSIZ];

	asprintf (buffer, ap);

	(void) fflush (stdout);

	fprintf (stderr, "%s: ", myname);
	(void) fputs (buffer, stderr);
	(void) fputc ('\n', stderr);

	(void) fflush (stderr);
}
#else
/* VARARGS */

void    advise (what, fmt)
char   *what,
       *fmt;
{
	advise (what, fmt);
}
#endif
